catwalk/models/rank_classification.py

```diff
@@ -178,11 +178,16 @@ def collate_for_training(self, instances: Sequence[Tuple[Task, Instance]]) -> An
             truncation=True,
             pad_to_multiple_of=8,
             return_tensors='pt',
-            is_split_into_words=False)
+            is_split_into_words=False,
+            return_token_type_ids=True,
+            )
         tokenized_strings['labels'] = torch.full_like(tokenized_strings.input_ids, -100)
         for i, label in enumerate(tokenized_strings.labels):
-            mask = [s == 1 for s in tokenized_strings.sequence_ids(i)]
+            mask = [bool(s == 1) for s in tokenized_strings['token_type_ids'][i]]
+            # have to account for any special tokens added to each of the two inputs
+            mask = [False] * self.tokenizer.num_special_tokens_to_add(pair=True) + mask
             label[mask] = tokenized_strings.input_ids[i, mask]
```

Comment:
I don't understand this. `mask` needs to be the same length as `input_ids`, right? You're making it longer though by adding `[False] * self.tokenizer.num_special_tokens_to_add(pair=True)` to the front? Also, that does not work when the special tokens go somewhere other than in front of the second sequence.

Was this always wrong?

Reply:
> I don't understand this. `mask` needs to be the same length as `input_ids`, right? You're making it longer though by adding `[False] * self.tokenizer.num_special_tokens_to_add(pair=True)` to the front? Also, that does not work when the special tokens go somewhere other than in front of the second sequence.
> 
> Was this always wrong?

So the weird thing is that unlike `.sequence_ids()`, `token_type_ids` does not account for special tokens that are added by the tokenizer. So when that is the case `tokenized_strings['token_type_ids'][I]` is actually shorter sequence length than `tokenized_strings['input_ids'][i]` by the amount of special tokens. This strikes me as a bug.

Padding with the `False` here is a work around. Yes I guess this does assume that special tokens aren't added somewhere else than the beginning cuz those wouldn't be counted. I couldn't find a convenient way to back out the information about where the special tokens go in the sequence, but I can take another look.